linux/devdoc/file_linux_requirements.md

```diff
@@ -0,0 +1,159 @@
+
+**SRS_FILE_LINUX_43_009: [** `file_read_async` shall initialize an `aiocb` struct with `handle` as `aio_fildes`, `position` as `aio_offset`, source as `aio_buf`, `source->alias->size` as `aio_nbytes`, and the initialized `sigevent` struct as `aio_sigevent`. **]**
+
+**SRS_FILE_LINUX_43_010: [** `file_read_async` shall call `aio_read` with the initialized `aiocb` struct as `aiocbp`.  **]**
+
+**SRS_FILE_LINUX_43_011: [** If `aio_read` fails, `file_read_async` shall return `FILE_READ_ASYNC_READ_ERROR`. **]**
+
+**SRS_FILE_LINUX_43_014: [** If `aio_read` succeeds, `file_read_async` shall return `FILE_READ_ASYNC_OK`. **]**
+
+**SRS_FILE_LINUX_43_015: [** If there are any failures, `file_read_async` shall return `FILE_READ_ASYNC_ERROR`. **]**
+
+
+## file_extend_filesize
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+**SRS_FILE_LINUX_43_018: [** `file_extend_filesize` shall return 0. **]**
+
+Will be implemented later.
+
+## on_file_io_complete_linux
+
+```c
+static void on_file_io_complete_linux( FILE_LINUX_IO* io);
+```
+
+`on_file_io_complete_linux` is called when a file operation completes asynchronously.
+
+**SRS_FILE_LINUX_43_021: [** `on_file_io_complete_linux` shall recover the `aiocb` struct that was used to create the current asynchronous operation. **]**
+
+**SRS_FILE_LINUX_43_022: [** `on_file_io_complete_linux` shall call `aio_return` to determine if the asynchronous operation succeeded. **]**
+
+**SRS_FILE_LINUX_43_023: [** If the asynchronous operation did not succeed, `on_file_io_complete_linux` shall call `user_callback` with `false` as `is_successful`. **]**
+
+**SRS_FILE_LINUX_43_024: [** If the type of the asynchronous operation is read, `on_file_io_complete_linux` shall  construct a `CONSTBUFFER_HANDLE` by calling `CONSTBUFFER_CreateWithMoveMemory` from the bytes read by `ReadFile`. **]**
+
+**SRS_FILE_LINUX_43_025: [** If the construction of the `CONSTBUFFER_HANDLE` fails, `on_file_io_complete_linux` shall call `user_callback` with `user_context`, `false` as `is_successful` and `NULL` as `content`. **]**
+
+**SRS_FILE_LINUX_43_026: [** If the construction of the `CONSTBUFFER_HANDLE` succeeds, `on_file_io_complete_linux` shall call `user_callback` with `user_context`, `true` as `is_successful` and the created `CONSTBUFFER_HANDLE` as `content`. **]**
+
+**SRS_FILE_LINUX_43_027: [** If the type of the asynchronous operation is write, `on_file_io_complete_linux` shall call `user_callback` with `user_context` and `true` as `is_successful`. **]**
+
+**SRS_FILE_LINUX_43_028: [** If the type of asynchronous operation is neither read nor write, `on_file_io_complete_win32` shall call `user_report_fault_callback` with `user_report_fault_context` which were specified when `file_create` was called. **]**

```

Comment:
>user_report_fault_callback [](start = 134, length = 26)

I think it should call user_report_fault_callback in absolutely all the cases where it fails in the callback. Ideally callbacks should have 0 (zero) possible places to fail, but that is sometimes not possible in practice.

Reply:
Hmm sounds more like something where we want an enum result in the user callback (read failed, "something else" failed). I think the report fault should only be called when we really don't have a context to call the read/write complete callback

---
In reply to: [450481659](https://github.com/Azure/azure-c-pal/pull/7#discussion_r450481659) [](ancestors = 450481659)

Reply:
the fault callback doesn't need to be called in linux in the next iteration. 

---
In reply to: [451690008](https://github.com/Azure/azure-c-pal/pull/7#discussion_r451690008) [](ancestors = 451690008,450481659)